### PR TITLE
Adds all() that accepts 4 heterogenous values

### LINKED ISSUE #61: all() with a generic tuple containing 4 types
The `all()` function currently supports tuples with up to [3 different types](https://github.com/google/promises/blob/1.2.3/Sources/Promises/Promise%2BAll.swift#L104). Does it make sense to have an `all()` function that supports tuples with up to 4 types? I realize that if 4 types are supported, then someone might ask for 5 types! So, I understand if this request is denied.

As a workaround, I'm composing `all` within `all`, which isn't pretty: `all(promise1, all(promise2, promise3, promise4))`. FYI, though I'm using 4 promises, my code rarely (if ever) makes 4 I/O requests due to caching.

### PR DESCRIPTION
**Pull request by EthanLozano**
Fixes #61

### CODE REVIEW

**Comment by EthanLozano** [comment]
I signed it!

**Comment by EthanLozano** [comment]
I'm not too proud of this code since it is just copy-pasted from the heterogeneous 3 version. But, it wasn't immediately obvious how I could have code re-use, so I just went with a quick and dirty solution.

File: Sources/Promises/Promise+All.swift [hunk 1]
```diff
@@ -133,3 +133,51 @@ public func all<A, B, C>(
...
+            let valueC = Promise<C>.asValue(values[2]),
+            let valueD = Promise<D>.asValue(values[3])
+      else {
+        preconditionFailure("Cannot convert \(type(of: objCValues)) to \((A, B, C, D).self)")
+      }
+      return (valueA, valueB, valueC, valueD)
+    })
+  )
+  // Keep Swift wrapper alive for chained promises until `ObjCPromise` counterpart is resolved.
+  promises.forEach {
```
Comment (EthanLozano) [inline Sources/Promises/Promise+All.swift]: we could probably get some code re-use here?

Comment (shoumikhin) [reply Sources/Promises/Promise+All.swift]: Should be fine for optimization purposes: no extra func call.

**Pull request merged**